internal/indexer/module_calls.go

```diff
@@ -11,104 +11,111 @@ import (
+		return jobIds, err
+	}
 
-		moduleCalls, err := idx.modStore.ModuleCalls(modHandle.Path())
+	var errs *multierror.Error
+
+	idx.logger.Printf("indexing installed module calls: %d", len(moduleCalls.Installed))
+	for _, mc := range moduleCalls.Installed {
+		fi, err := os.Stat(mc.Path)
+		if err != nil || !fi.IsDir() {
+			multierror.Append(errs, err)
+			continue
+		}
+		err = idx.modStore.Add(mc.Path)
 		if err != nil {
-			return jobIds, err
+			multierror.Append(errs, err)
+			continue
 		}
 
-		var errs *multierror.Error
+		mcHandle := document.DirHandleFromPath(mc.Path)
+		// copy path for queued jobs below
+		mcPath := mc.Path
 
-		for _, mc := range moduleCalls.Installed {
-			fi, err := os.Stat(mc.Path)
-			if err != nil || !fi.IsDir() {
-				multierror.Append(errs, err)
-				continue
-			}
-			err = idx.modStore.Add(mc.Path)
-			if err != nil {
-				multierror.Append(errs, err)
-				continue
-			}
+		refCollectionDeps := make(job.IDs, 0)
 
-			mcHandle := document.DirHandleFromPath(mc.Path)
-			// copy path for queued jobs below
-			mcPath := mc.Path
+		parseId, err := idx.jobStore.EnqueueJob(job.Job{
+			Dir: mcHandle,
+			Func: func(ctx context.Context) error {
+				return module.ParseModuleConfiguration(idx.fs, idx.modStore, mcPath)
+			},
+			Type: op.OpTypeParseModuleConfiguration.String(),
+		})
+		if err != nil {
+			multierror.Append(errs, err)
+		} else {
+			jobIds = append(jobIds, parseId)
+			refCollectionDeps = append(refCollectionDeps, parseId)
+		}
 
-			id, err := idx.jobStore.EnqueueJob(job.Job{
-				Dir: mcHandle,
+		var metaId job.ID
+		if parseId != "" {
+			metaId, err = idx.jobStore.EnqueueJob(job.Job{
+				Dir:  mcHandle,
+				Type: op.OpTypeLoadModuleMetadata.String(),
 				Func: func(ctx context.Context) error {
-					return module.ParseModuleConfiguration(idx.fs, idx.modStore, mcPath)
-				},
-				Type: op.OpTypeParseModuleConfiguration.String(),
-				Defer: func(ctx context.Context, jobErr error) (job.IDs, error) {
-					ids := make(job.IDs, 0)
-
-					id, err := idx.jobStore.EnqueueJob(job.Job{
-						Dir:  mcHandle,
-						Type: op.OpTypeLoadModuleMetadata.String(),
-						Func: func(ctx context.Context) error {
-							return module.LoadModuleMetadata(idx.modStore, mcPath)
-						},
-						Defer: func(ctx context.Context, jobErr error) (job.IDs, error) {
-							return idx.collectReferences(ctx, mcHandle)
-						},
-					})
-					if err != nil {
-						return ids, err
-					} else {
-						ids = append(ids, id)
-					}
-
-					return ids, nil
+					return module.LoadModuleMetadata(idx.modStore, mcPath)
 				},
+				DependsOn: job.IDs{parseId},
 			})
 			if err != nil {
 				multierror.Append(errs, err)
-				continue
+			} else {
+				jobIds = append(jobIds, metaId)
+				refCollectionDeps = append(refCollectionDeps, metaId)
 			}
-			jobIds = append(jobIds, id)
+		}
 
-			id, err = idx.jobStore.EnqueueJob(job.Job{
+		if parseId != "" {
+			ids, err := idx.collectReferences(mcHandle, refCollectionDeps)
```

Comment:
I didn't notice this before: why are we calling `collectReferences` here, before parsing variables? While in the walker, we're calling it much later?

Reply:
The ordering of _scheduling_ is not significant - the `*.tfvars` parsing and `*.tf` references should be independent of each other so it doesn't matter in what order do they run - we just happen to schedule them in different order in the two places, `DependsOn` and `Defer` (if any) decide in what order they actually get executed.

Reply:
Okay! Thanks for the explanation!

Reply:
We could align the ordering of scheduling, but there are also other differences - e.g. that when we parse submodules here, we assume that the modules are not initialised by themselves (given the context) and don't attempt to parse module manifest or obtain schema - both of which would otherwise be two more jobs which reference collection should depend on.